core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java

```diff
@@ -0,0 +1,291 @@
+            currentObj = it.getValue();
+            nonNestedField += "." + it.getKey();
+          } else if (currentObj instanceof ExprCollectionValue) {
+            currentObj = currentObj.collectionValue().get(0);
+          } else {
+            nestingComplete = true;
+          }
+        }
+        this.nonNestedFields.add(nonNestedField);
+      }
+    }
+  }
+
+
+  /**
+   * Simplifies the structure of row's source Map by flattening it,
+   * making the full path of an object the key
+   * and the Object it refers to the value.
+   *
+   * <p>Sample input:
+   * keys = ['comments.likes']
+   * row = comments: {
+   * likes: 2
+   * }
+   *
+   * <p>Return:
+   * flattenedRow = {comment.likes: 2}
+   *
+   * @param nestedField : Field to query in row.
+   * @param row : Row returned from OS.
+   * @param prevList : List of previous nested calls.
+   * @param supportArrays : When false we do not need to execute a cross join.
+   * @return : List of nested select items or cartesian product of nested calls.
+   */
+  private List<Map<String, ExprValue>> flatten(
+      String nestedField,
+      ExprValue row,
+      List<Map<String,
+      ExprValue>> prevList,
+      boolean supportArrays
+  ) {
+    List<Map<String, ExprValue>> copy = new ArrayList<>();
+    List<Map<String, ExprValue>> newList = new ArrayList<>();
+
+    ExprValue nestedObj = null;
+    getNested(nestedField, nestedField, row, copy, nestedObj, supportArrays);
+
+    // Only one field in select statement
+    if (prevList.size() == 0) {
+      return copy;
+    }
+
+    if (!supportArrays || containSamePath(copy.get(0))) {
+      var resultIt = this.result.iterator();
+      Map<String, ExprValue> resultVal = resultIt.next();
+      var copyIt = copy.iterator();
+      Map<String, ExprValue> copyVal = copyIt.next();
+      for (int i = 0; i < this.result.size(); i++) {
+        resultVal.putAll(copyVal);
+        if (copyIt.hasNext()) {
+          copyVal = copyIt.next();
+        }
+        if (resultIt.hasNext()) {
+          resultVal = resultIt.next();
+        }
+      }
+      return this.result;
+    } else {
+      // Generate cartesian product
+      for (Map<String, ExprValue> prevMap : prevList) {
+        for (Map<String, ExprValue> newMap : copy) {
+          newList.add(Stream.of(newMap, prevMap)
+              .flatMap(map -> map.entrySet().stream())
+              .collect(Collectors.toMap(
+                  Map.Entry::getKey,
+                  Map.Entry::getValue)));
+        }
+      }
+      return newList;
+    }
+  }
+
+  /**
+   * Check if newMap field has any sharing paths in prevMap.
+   * @param newMap : New map to add to result set.
+   * @return : true if there is already a field added to result set with same path.
+   */
+  boolean containSamePath(Map<String, ExprValue> newMap) {
+    String newKey = newMap.keySet().iterator().next();
+    Map<String, ExprValue> resultMap = this.result.iterator().next();
```

Comment:
You check only first entry in `result`. Why?

Reply:
In the event we are adding a nested field with another fields same path to the result set, we don't want to do a cross join on them. So we check which paths are already added to the results map to compare with the new map being added. Even if a previous item was `null` or `missing` their key will still be in the results map. Therefore we only care to check the first entry of `result`.